### PR TITLE
swarm: Add 'swarm ls' to list manifest directory trees

### PR DESCRIPTION
**Pull request by lmars**
This adds a `swarm ls` command which lists files and directories stored in a manifest.

Rather than listing all files, it uses "directory prefixes" in case there are a lot of files in a manifest but you just want to traverse it.

This also includes some refactoring to the tests and the introduction of a `swarm/api/client` package to make things easier to test.

Examples:

---

1. with the contents of the `swarm` directory of this repository uploaded with manifest hash `47203951b3451bcb73f2a3764142e9b2987c855764a08dc3bdd11b1e798b3b04`:

```
$ swarm ls 47203951b3451bcb73f2a3764142e9b2987c855764a08dc3bdd11b1e798b3b04
HASH                                                              CONTENT TYPE  PATH
                                                                  DIR           api/
                                                                  DIR           network/
                                                                  DIR           services/
                                                                  DIR           storage/
27d7a6d01d1f6a6e143f368a815352205c4b9b5bf4c3efa51f04bfc2d675f971                swarm.go
                                                                  DIR           testutil/
```

```
$ swarm ls 47203951b3451bcb73f2a3764142e9b2987c855764a08dc3bdd11b1e798b3b04 services/
HASH  CONTENT TYPE  PATH
      DIR           services/swap/
```

```
$ swarm ls 47203951b3451bcb73f2a3764142e9b2987c855764a08dc3bdd11b1e798b3b04 services/swap/
HASH                                                              CONTENT TYPE  PATH
8cdf723c59f002addc37ec1f6873403ab9532117f37aa385412c8c89bd0a0c28                services/swap/swap.go
                                                                  DIR           services/swap/swap/
```

---

2. listing files in the public `theswarm.eth` manifest:

```
$ swarm --bzzapi "http://swarm-gateways.net" ls theswarm.eth
HASH                                                              CONTENT TYPE               PATH
                                                                  DIR                        .git/
3f765acb4c6a9645d27fa54532926297955da67fb8a088026d413be7f6125e25  text/plain; charset=utf-8  README.md
                                                                  DIR                        Swarm_files/
                                                                  DIR                        ethersphere/
                                                                  DIR                        images/
7b7eb0f53ef673191338c4c6ecd8997a8016397cd1c3f2774dc238840c9e5f0f  text/html; charset=utf-8   index.html
                                                                  DIR                        talks/
7b7eb0f53ef673191338c4c6ecd8997a8016397cd1c3f2774dc238840c9e5f0f  text/html; charset=utf-8   /
```

```
$ swarm --bzzapi "http://swarm-gateways.net" ls theswarm.eth images/
HASH                                                              CONTENT TYPE  PATH
736f17d25c07a0a1074b838e90bde4f9cfaeeb940e3ef545ca355eecb96630e5  image/png     images/swarm-01.png
ffc0988c581f0fcbb658f66b5ea636e63abf746ecb717931862b8e6cca3e746d  image/png     images/swarm-02.png
a17efc88ece60835374b91680af4f7de2042dab54c0f638eb7bfc2f53691585c  image/png     images/swarm-03.png
b645405800c2f2716c92e606eb217d566257754e484c55f19194f5c9badc43a6  image/png     images/swarm-04.png
```

---

Ref #3540 (does not close it as I haven't added a `--json` flag).

/cc @zelig

### CODE REVIEW

File: cmd/swarm/list.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+	if len(args) < 1 {
+		utils.Fatalf("Please supply a manifest reference as the first argument")
+	}
+	manifest := args[0]
+
+	var prefix string
+	if len(args) == 2 {
+		prefix = args[1]
+	}
```
Comment (zelig) [inline cmd/swarm/list.go]: 3, 4 arguments are silently ignored. Maybe warn

**Comment by lmars** [comment]
@zelig extra arguments are now handled (I do feel like the CLI library should be handling that, I am used to [docopt](http://docopt.org/) where the usage is used to generate a parser so you don't have to :smile:).

**Comment by zelig** [comment]
@karalabe I wonder what these timouts are https://circleci.com/gh/ethereum/go-ethereum/2595?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by zelig** [comment]
reviewed @lmars and only the CI test issues outstanding 
I know this is superceded by https://github.com/ethereum/go-ethereum/issues/3773
but if that does not make it, lets make sure this goes through

**Comment by lmars** [comment]
@zelig I have fixed the CI issue, feel free to merge and I will rebase #3773 on top.

**Pull request merged**